SystemMicrocontroller_Sensors/Core/Src/lcd20x4.c

```diff
@@ -0,0 +1,331 @@
+	CurrentGPIO = GPIOG;
+	if(isCurrentGPIOUsed){
+		__HAL_RCC_GPIOG_CLK_ENABLE();
+	}
+	CurrentGPIO = GPIOH;
+	if(isCurrentGPIOUsed){
+		__HAL_RCC_GPIOH_CLK_ENABLE();
+	}
+
+}
+
+//Prints to current cursor position
+void LCD_Printf(LCD_HandleTypeDef * lcd, const char* str, ...)
+{
+	char stringArray[80];
+	va_list args;
+	va_start(args, str);
+	vsprintf(stringArray, str, args);
+	va_end(args);
+	for(uint8_t i=0;  i<strlen(stringArray) && i<80; i++)
+	{
+		LCD_WriteData(lcd, (uint8_t)stringArray[i]);
+	}
+}
+
+void LCD_PrintTo1stLine(LCD_HandleTypeDef * lcd, const char* str, ...)
+{
+	char stringArray[20];
+	va_list args;
+	va_start(args, str);
+	vsprintf(stringArray, str, args);
+	va_end(args);
+
+	LCD_SetCursorPosition(lcd, 0, 0);
+	for(uint8_t i=0;  i<strlen(stringArray) && i<20; i++)
+	{
+		LCD_WriteData(lcd, (uint8_t)stringArray[i]);
+	}
+}
+
+void LCD_PrintTo2ndLine(LCD_HandleTypeDef * lcd, const char* str, ...)
+{
+	char stringArray[20];
+	va_list args;
+	va_start(args, str);
+	vsprintf(stringArray, str, args);
+	va_end(args);
+
+	LCD_SetCursorPosition(lcd, 1, 0);
+	for(uint8_t i=0;  i<strlen(stringArray) && i<20; i++)
+	{
+		LCD_WriteData(lcd, (uint8_t)stringArray[i]);
+	}
+}
+
+void LCD_PrintTo3rdLine(LCD_HandleTypeDef * lcd, const char* str, ...)
+{
+	char stringArray[20];
+	va_list args;
+	va_start(args, str);
+	vsprintf(stringArray, str, args);
+	va_end(args);
+
+	LCD_SetCursorPosition(lcd, 2, 0);
+	for(uint8_t i=0;  i<strlen(stringArray) && i<20; i++)
+	{
+		LCD_WriteData(lcd, (uint8_t)stringArray[i]);
+	}
+}
+
+void LCD_PrintTo4thLine(LCD_HandleTypeDef * lcd, const char* str, ...)
+{
+	char stringArray[20];
+	va_list args;
+	va_start(args, str);
+	vsprintf(stringArray, str, args);
+	va_end(args);
+
+	LCD_SetCursorPosition(lcd, 3, 0);
+	for(uint8_t i=0;  i<strlen(stringArray) && i<20; i++)
+	{
+		LCD_WriteData(lcd, (uint8_t)stringArray[i]);
+	}
+}
+
+
+/*
+ * CursorMode may take any of the following values:
+ * LCD_CURSOR_OFF
+ * LCD_CURSOR_ON_BLINK_OFF
+ * LCD_CURSOR_ON_BLINK_ON
+ */
+void LCD_SetCursorMode(LCD_HandleTypeDef * lcd, uint8_t CursorMode)
+{
+  if(CursorMode == LCD_CURSOR_ON_BLINK_ON){
+	  LCD_WriteCommand(lcd, COMMAND_DISPLAY_ON_OFF_CONTROL | OPT_D | OPT_C | OPT_B);
+	  lcd->CursorMode = LCD_CURSOR_ON_BLINK_ON;
+  }
+  else if(CursorMode == LCD_CURSOR_ON_BLINK_OFF){
+	  LCD_WriteCommand(lcd, COMMAND_DISPLAY_ON_OFF_CONTROL | OPT_D | OPT_C);
+	  lcd->CursorMode = LCD_CURSOR_ON_BLINK_OFF;
+  }
+  else if(CursorMode == LCD_CURSOR_OFF){
+	  LCD_WriteCommand(lcd, COMMAND_DISPLAY_ON_OFF_CONTROL | OPT_D);
+	  lcd->CursorMode = LCD_CURSOR_OFF;
+  }
+}
+
+void LCD_SetCursorPosition(LCD_HandleTypeDef * lcd, uint8_t row, uint8_t col)
+{
+	LCD_WriteCommand(lcd, COMMAND_SET_DDRAM_ADDR + ROW_20[row] + col);
+}
+
+
+void LCD_Clear(LCD_HandleTypeDef * lcd){
+	LCD_WriteCommand(lcd, COMMAND_CLEAR_DISPLAY);
+	DELAY(1);
+}
+
+/**
+ * Write a byte to the command register
+ */
+void LCD_WriteCommand(LCD_HandleTypeDef * lcd, uint8_t command)
```

Comment:
May this function be called from outside the driver? Internal function must be protected with keyword static.